src/array/cpu/concurrent_id_hash_map.h

```diff
@@ -0,0 +1,205 @@
+    bool Contains(IdType id) const {
+        IdType j;
+        for (j = (id&mask); hmap[j].key != -1 && hmap[j].key != id; j = ((j + 1) & mask)) {}
+        return hmap[j].key != -1;
+    }
+
+    // Return the new id of the given id. If the given id is not contained
+    // in the hash map, returns the default_val instead.
+    IdType Map(IdType id, IdType default_val) const {
+        IdType pos = (id & mask);
+        IdType delta = 10;
+        while (hmap[pos].key != -1 && hmap[pos].key != id) {
+            pos = ((pos + delta*delta) & mask);
+            delta += 1;
+        }
+
+        if (hmap[pos].key == -1) {
+            return default_val;
+        } else {
+            return hmap[pos].value;
+        }
+    }
+
+    void Map(IdArray ids, IdType default_val, IdArray& new_ids) const {
+        const IdType* ids_data = ids.Ptr<IdType>();
+        const size_t len = static_cast<size_t>(ids->shape[0]);
+        IdType* values_data = new_ids.Ptr<IdType>();
+
+        size_t thread_num = omp_get_max_threads();
+        int block_size = (len + thread_num - 1) / thread_num;
+        #pragma omp parallel
+        {
+        #pragma omp for
+            for (size_t i = 0; i < thread_num; ++i) {
+                size_t start = i *block_size;
+                size_t end = (i + 1) *block_size;
+                end = std::min(end, len);
+
+                for (size_t index = start; index < end; ++index) {
+                    values_data[index] = Map(ids_data[index], default_val);
+                }
+            }
+        }
+    }
+
+    ~ConcurrentIdHashMap() {
+        if (hmap != NULL) {
+            DGLContext ctx;
+            ctx.device_type = kDGLCPU;
+            auto device = runtime::DeviceAPI::Get(ctx);
+            device->FreeWorkspace(ctx, hmap);
+        }
+    }
+
+    void insert_cas(IdType id, std::vector<bool>& valid, size_t index) {
+        IdType pos = (id & mask);
+        IdType delta = 10;
+        while (!attempt_insert_at(pos, id, valid, index)) {
+            pos = ((pos + delta*delta) & mask);
+            delta += 1;
+        }
+    }
+  
+    // Key must exist.
+    void set_value(IdType k, IdType v) {
+        IdType pos = (k & mask), delta = 10;
+        while (hmap[pos].key != k) {
+            pos = ((pos + delta*delta) & mask);
+            delta+=1;
+        }
+        
+        hmap[pos].value = v;
+    }
+
+    bool attempt_insert_at(int64_t pos, IdType key, std::vector<bool>& valid, size_t index) {
+        IdType a = kEmptyKey;
+        bool success = hmap[pos].key.compare_exchange_weak(a, key);
+        if (success) {
+            valid[index] = true;
+            return true;
+        }
+        else {
+            if(hmap[pos].key == key)
+                return true;
+            else
+                return false; 
+        }
+    }
+
+ protected:
+    static constexpr IdType kEmptyKey = static_cast<IdType>(-1);
+    Mapping* hmap;
+    IdType mask;
```

Comment:
The protected/private members (`hmap` and `mask`) are better suffixed with `_` to distinguish from local variables.

Reply:
fixed